include/aws/io/host_resolver.h

```diff
@@ -211,6 +223,84 @@ AWS_IO_API size_t aws_host_resolver_get_host_address_count(
     const struct aws_string *host_name,
     uint32_t flags);
 
+struct aws_host_resolver_listener;
+
+struct aws_host_resolver_listener_vtable {
+    void (*acquire)(struct aws_host_resolver_listener *listener);
+    void (*release)(struct aws_host_resolver_listener *listener);
+};
+
+struct aws_host_resolver_listener {
+    struct aws_host_resolver_listener_vtable *vtable;
+    void *impl;
+};
+
+/* Callback for receiving a new host addresses from a listener. Memory for host_address is only guaranteed to exist
+ * during the callback, and must be copied if the caller needs it to persist after. */
+typedef void(aws_host_resolver_resolved_address_fn)(
+    struct aws_host_resolver_listener *listener,
+    struct aws_host_address *host_address,
+    void *user_data);
+
+typedef void(aws_host_resolver_listener_shutdown_fn)(void *user_data);
+
+struct aws_host_resolver_listener_options {
+
+    /* Name of the host to listen for notifications from. */
+    const struct aws_string *host_name;
+
+    /* Callback for when an address is resolved for the specified host. */
+    aws_host_resolver_resolved_address_fn *resolved_address_callback;
```

Comment:
Not sure this new functionality belongs on the base class vtable. It seems specific to the `aws_100gb_nic_soaking_s3_wet_dream_host_resolver` which aggressively pre-caches addresses. Like, if I were to implement a power-conscious host resolver for use by IoT, it would only resolve hosts when asked to, and a callback like this that tells you about unsolicited background activity would make no sense.

But seriously, maybe today's the day we rename the "default" host resolver to `aws_host_resolver_s3_performance` or something like that.

Reply:
I was thinking listener functionality would be optionally overridable, which is how I implemented it in the source file, but I'm realizing I should have commented that as such in the header file.  (Not sure that that solves your concerns though.)